cyclonedx/model/component.py

```diff
@@ -1783,7 +1783,7 @@ def __hash__(self) -> int:
             self.mime_type, self.supplier, self.author, self.publisher,
             self.description, self.scope, tuple(self.hashes),
             tuple(self.licenses), self.copyright, self.cpe,
-            self.purl,
+            self.purl, self.bom_ref.value,
```

Comment:
I see why you go with `self.bom_ref.value` instead of `self.bom_ref`.
`BomRef.__hash__` exists and is implemented to account for the `__id__` of the "empty" object entity, so that `None` values are unequal.
Your implementation does not do so, and may cause unexpected behaviour here.

Reply:
Indeed I had originally only added `self.bom_ref`, but that would then _not_ deduplicate the `None`s. Is there a reason for the `BomRef` hash to include the id? It seems superfluous when there is a unique identifier. Note also that the code claims a random UUIDv4 should have been assigned when there is no ref, but this is not actually the case